core/src/actors/codec.rs

```diff
@@ -29,21 +45,17 @@ impl Decoder for P2PCodec {
 
     /// Method to decode bytes to a request
     fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
-        // Locate a byte corresponding to a '\n' in the byte stream
-        if let Some(i) = src.iter().position(|&b| b == b'\n') {
-            // Remove the serialized frame from the buffer.
-            let line = src.split_to(i + 1);
-
-            // Parse the buffer as an UTF-8 encoded string
-            let mut res = String::from_utf8(line.to_vec()).unwrap();
-
-            // Remove the last two bytes of the string (corresponding to \r\n)
-            res.truncate(res.len() - 2);
-
-            Ok(Some(Request::Message(res)))
-        } else {
-            Ok(None)
+        let mut ftb: Option<Self::Item> = None;
+        let msg_len = src.len();
+        if msg_len > 2 {
+            let mut header_vec = Cursor::new(&src[0..2]);
+            let msg_size = header_vec.read_u16::<BigEndian>().unwrap() as usize;
+            if msg_len >= msg_size + HEADER_SIZE {
+                src.split_to(HEADER_SIZE);
+                ftb = Some(Request::Message(src.split_to(msg_size)));
+            }
         }
+        Ok(ftb)
```

Comment:
I liked it! Perhaps, I would include some comments of what is happening in case the *if* condition is not fulfilled. This would help to understand the code if future changes are needed. :relaxed: